### ISSUE TITLE
Abstract algorithms?

### ISSUE THREAD

**Issue by santiagogil** [flumedb/flumedb#4]
Hi Dominic.
I fallen in love instantly with this project and started to look for things to make into flume views.

For each algorithm, let's say a bloom filter, there's a lot of diferent existing and posible implementations. I find this can lead to a cluttered ecosystem around flumedb with lots of flumeview bloom filters with slightly different names and flumeview specific code.

What if we make an abstract contrat of  each algorithm and a corresponding flumeview wich accepts any compatible implementation.

Then:
+  We can have just one flumeview-bloom, flumeview-hashtable, etc.
+  It would be trivial to wrap any implementation to make it compatible.
+  Users can swap implementations easily without writing any flumeview specific code.
+  Other projects can benefit of the unified contract of each used algorithm.

Does it make sence?

**Comment by dominictarr** [comment]
Yes! this is great, although, since each flumeview is quite narrowly focused on a specific feature, I think it's pretty easy to wrap one to another - but at the same time, it's easy to give them the same api. We just have to make sure that the flume community is a community that talks to each other ;)

I'd rather not have to import a thing to inherit from something to create a specific type of view (it's especially painful when it's the first implementation of a type).

In javascript land, I think the best way to do this is to have reusable test cases.
we have this for flumelog already https://github.com/flumedb/test-flumelog

And this pattern of reusable tests is the important part of http://github.com/level/abstract-leveldown
and I also use to test chloride https://github.com/dominictarr/chloride-test (and the various browser polyfils)

**Comment by santiagogil** [comment]
Yes. I was thinking about reusable test cases for algorithm apis.

Let's say we have a module flumeview-bloom that accepts as parameter any bloom filter that is abstract-bloom compatible.

I know it's not the fastest way for the initial implementation, but each view needs preety different apis (bloom vs trie for example).

What I was looking for was a way to swap just the algorithm implementation without changing any flumeview code.

With an abstract-flumeview we can test that a module is a proper flumeview but not the specific methods that each algorithm needs.

A few minutes later found myself with 20 tabs open trying to find any usefull generalization over trie implementations just to realize that I was overstandarizing :D

So, let's stay with an abstract-flumeview that does not try to impose anything to algorithm implementations.

Sorry about the premature "what if..."
I'm starting to think is a compulsion :(

**Comment by dominictarr** [comment]
the trouble is that abstract-flumeview is really just an interface - it doesn't have anything that you can really test. it's basically just this:

``` js
module.exports = function createFumeView(...) {
  return function (log, name) {
    var since = observable()
    //read where the current persisted state is up to, or to start over
    //set since to -1.

    return {
      since: since,
      methods: {}, //describe the methods your flumeview adds
      //process incoming data as a sink pull-stream
      createSink: createSinkStream(cb),
      //expose the methods your flumeview exports...
    }
  }
}
```

it's just a pattern, there isn't very much that is common to all flumeviews.
(although, any particular algorithm - all tries or all bloom filters, etc, quite a bit)

**Comment by christianbundy** [comment]
If we had a wrapper, it might be useful to add the pull-write code so that we don't have to implement things like batching in each view. I'd imagine that there are probably other common threads, like settings `since`, but I'm not sure how much easier it could even be.

**Comment by dominictarr** [comment]
I think helper libraries are the answer here. You could make a module that makes it easy to create a view, and refactor the current views to use it. But I suspect that different views have their own needs that there isn't very much overlap. especially when you get to stuff like: https://github.com/flumedb/flumeview-vector
